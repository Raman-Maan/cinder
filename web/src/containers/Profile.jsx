import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  Container,
  ModalHeader,
  ModalBody,
  ModalFooter,
  Button,
  Input
} from 'reactstrap';
import { fetchFilters, hideProfile } from '../actions';
import FilterElement from '../components/FilterElement';
import Dropdown from '../components/Dropdown';

import './styles/Profile.css';

export class Profile extends Component {
  constructor(props) {
    super(props);

    this.toggle = this.toggle.bind(this);

    this.state = {
      filters: {
        gender: null
      }
    };
  }

  // Call this function to fetch filters
  // id and token are in this.props.userID and this.props.token
  // filters will be in this.props.filters
  fetchFilters(userID, token) {
    this.props.fetchFilters(userID, token);
  }

  toggle(e) {
    e.stopPropagation();
    this.props.hideProfile();
  }

  onElementToggle(field, value) {
    this.setState(prev => ({
      ...prev,
      filters: {
        ...prev.filters,
        [field]: {
          ...prev.filters[field],
          active: value
        }
      }
    }));
  }

  onDropdownChange(field, value) {
    this.setState(prev => ({
      ...prev,
      filters: {
        ...prev.filters,
        [field]: {
          ...prev.filters[field],
          values: value
        }
      }
    }));
  }

  renderBody() {
    return (
      <div>
        <div className="name">
          <h5>Display Name</h5>
          <Input type="text" defaultValue={this.props.userInfo.userName} />
        </div>
        <hr />
        <div className="birthday">
          <h5>Birthday</h5>
          <Input
            type="date"
            defaultValue={
              new Date(this.props.userInfo.birthday).toJSON().split('T')[0]
            }
          />
        </div>
        <hr />
        <div className="bio">
          <h5>User Bio</h5>
          <Input type="textarea" defaultValue={this.props.userInfo.userBio} />
        </div>
        <hr />
        <div className="filters">
          <h5>User Filters</h5>
          <div className="gender">
            <h6>Gender Filter</h6>
            <FilterElement
              round
              onChange={this.onElementToggle.bind(this, 'gender')}
            >
              <Dropdown
                token={this.props.token}
                endpoint="/api/ref/gender"
                onChange={this.onDropdownChange.bind(this, 'gender')}
              />
            </FilterElement>
          </div>
        </div>
      </div>
    );
  }

  render() {
    const modal = !this.props.isVisible ? null : (
      <div className="profile modal" onClick={e => this.toggle(e)}>
        <Container>
          <div className="modal-content" onClick={e => e.stopPropagation()}>
            <ModalHeader toggle={e => this.toggle(e)}>My Profile</ModalHeader>
            <ModalBody>{this.renderBody()}</ModalBody>
            <ModalFooter>
              <Button color="secondary" outline onClick={e => this.toggle(e)}>
                Discard Changes
              </Button>
              <Button color="primary">Save Changes</Button>
            </ModalFooter>
          </div>
        </Container>
      </div>
    );

    return <div>{modal}</div>;
  }
}

const mapStateToProps = state => ({
  isVisible: state.profileDisplay.isVisible,
  userID: state.auth.userID,
  token: state.auth.token,
  filters: state.filters.filters
});

const mapDispatchToProps = dispatch => ({
  fetchFilters: (userID, token) => dispatch(fetchFilters(userID, token)),
  hideProfile: () => dispatch(hideProfile())
});

export default connect(mapStateToProps, mapDispatchToProps)(Profile);
