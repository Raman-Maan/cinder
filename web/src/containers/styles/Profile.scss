@import '../../assets/bootstrap/overrides/variables';

.profile.modal {
  display: block;
  background-color: $profile-modal-bg-color;

  .container {
    height: 100%;
    padding-top: 2em;
    padding-bottom: 2em;

    .modal-content {
      max-height: 100%;
      min-height: 50%;

      .modal-body {
        overflow: auto;
      }
    }
  }
}
