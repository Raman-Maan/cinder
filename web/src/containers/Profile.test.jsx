import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import fetchMock from 'fetch-mock';
import sinon from 'sinon';

import { Profile } from './Profile';

describe('<Profile />', () => {
  const testProfile = {
    Bio: 'This is a test bio for a test user',
    Birthday: '1991-10-18T05:00:00.000Z',
    UserName: 'Test User'
  };

  describe('Snapshot tests', () => {
    it('should be hidden if view is false', () => {
      const component = renderer.create(<Profile fetchData={jest.fn()} />);
      expect(component).toMatchSnapshot();
    });

    it('should display if profile properties are all empty', () => {
      const component = renderer.create(
        <Profile
          view
          fetchData={jest.fn()}
          profile={{ Bio: null, Birthday: null, UserName: null }}
        />
      );
      expect(component).toMatchSnapshot();
    });

    it('should display profile properties by default', () => {
      const component = renderer.create(
        <Profile view fetchData={jest.fn()} profile={testProfile} />
      );
      expect(component).toMatchSnapshot();
    });
  });

  describe('Enzyme tests', () => {
    let wrapper;
    let mockFetchData;
    let mockClickHandler;

    beforeEach(() => {
      mockFetchData = jest.fn();
      mockClickHandler = jest.fn();
      wrapper = mount(
        <Profile
          view
          profile={testProfile}
          fetchData={mockFetchData}
          clickHandler={mockClickHandler}
        />
      );
    });

    it('should fetch profile data on mount', () => {
      expect(mockFetchData.mock.calls.length).toBe(1);
    });

    describe('Gender Filter', () => {
      let genderFilter;
      let defaultState;

      fetchMock.get(/\/api\/ref\/gender/, [
        { id: 1, value: 'test1' },
        { id: 2, value: 'test2' },
        { id: 3, value: 'test3' }
      ]);

      beforeEach(() => {
        fetchMock.reset();
        genderFilter = wrapper.find('div.filters .gender');
        defaultState = wrapper.state();
      });

      it('should update state on toggle', () => {
        expect(wrapper.find('.filter.element.disabled')).toHaveLength(1);
        expect(!!defaultState.filters).toEqual(true);
        expect(defaultState.filters.gender).toBe(null);

        genderFilter.find('.toggle input').simulate('change');

        expect(wrapper.find('.filter.element.disabled')).toHaveLength(0);
        expect(!!wrapper.state().filters).toEqual(true);
        expect(wrapper.state().filters.gender).toEqual({ active: true });

        genderFilter.find('.toggle input').simulate('change');

        expect(wrapper.find('.filter.element.disabled')).toHaveLength(1);
        expect(!!wrapper.state().filters).toEqual(true);
        expect(wrapper.state().filters.gender).toEqual({ active: false });
      });
    });
  });
});
