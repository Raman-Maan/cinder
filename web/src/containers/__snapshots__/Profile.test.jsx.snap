// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> Snapshot tests should be hidden if view is false 1`] = `<div />`;

exports[`<Profile /> Snapshot tests should display if profile properties are all empty 1`] = `
<div>
  <div
    className="profile modal"
    onClick={[Function]}
  >
    <div
      className="container"
    >
      <div
        className="modal-content"
        onClick={[Function]}
      >
        <div
          className="modal-header"
        >
          <h4
            className="modal-title"
          >
            My Profile
          </h4>
          <button
            aria-label="Close"
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <div>
            <div
              className="name"
            >
              <h5>
                Display Name
              </h5>
              <input
                className="form-control"
                defaultValue={null}
                type="text"
              />
            </div>
            <hr />
            <div
              className="birthday"
            >
              <h5>
                Birthday
              </h5>
              <input
                className="form-control"
                defaultValue="1970-01-01"
                type="date"
              />
            </div>
            <hr />
            <div
              className="bio"
            >
              <h5>
                User Bio
              </h5>
              <textarea
                className="form-control"
                defaultValue={null}
              />
            </div>
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-outline-secondary"
            onClick={[Function]}
            type="button"
          >
            Discard Changes
          </button>
          <button
            className="btn btn-primary"
            onClick={[Function]}
            type={undefined}
          >
            Save Changes
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Profile /> Snapshot tests should display profile properties by default 1`] = `
<div>
  <div
    className="profile modal"
    onClick={[Function]}
  >
    <div
      className="container"
    >
      <div
        className="modal-content"
        onClick={[Function]}
      >
        <div
          className="modal-header"
        >
          <h4
            className="modal-title"
          >
            My Profile
          </h4>
          <button
            aria-label="Close"
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <div>
            <div
              className="name"
            >
              <h5>
                Display Name
              </h5>
              <input
                className="form-control"
                defaultValue="Test User"
                type="text"
              />
            </div>
            <hr />
            <div
              className="birthday"
            >
              <h5>
                Birthday
              </h5>
              <input
                className="form-control"
                defaultValue="1991-10-18"
                type="date"
              />
            </div>
            <hr />
            <div
              className="bio"
            >
              <h5>
                User Bio
              </h5>
              <textarea
                className="form-control"
                defaultValue="This is a test bio for a test user"
              />
            </div>
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-outline-secondary"
            onClick={[Function]}
            type="button"
          >
            Discard Changes
          </button>
          <button
            className="btn btn-primary"
            onClick={[Function]}
            type={undefined}
          >
            Save Changes
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
